Framework/Algorithms/test/CreateMonteCarloWorkspaceTest.h

```diff
@@ -0,0 +1,152 @@
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAlgorithms/CreateMonteCarloWorkspace.h"
+#include "MantidDataObjects/Workspace2D.h"
+#include "MantidFrameworkTestHelpers/WorkspaceCreationHelper.h"
+#include "MantidHistogramData/Histogram.h"
+#include <cxxtest/TestSuite.h>
+
+#include <random>
+
+using namespace Mantid::Algorithms;
+using namespace Mantid::API;
+using namespace Mantid::HistogramData;
+
+using Mantid::Algorithms::CreateMonteCarloWorkspace;
+
+class CreateMonteCarloWorkspaceTest : public CxxTest::TestSuite {
+public:
+  static CreateMonteCarloWorkspaceTest *createSuite() { return new CreateMonteCarloWorkspaceTest(); }
+  static void destroySuite(CreateMonteCarloWorkspaceTest *suite) { delete suite; }
+
+  void test_Init() {
+    CreateMonteCarloWorkspace alg;
+    TS_ASSERT_THROWS_NOTHING(alg.initialize());
+    TS_ASSERT(alg.isInitialized());
+  }
+
+  void test_computeNumberOfIterations() {
+    CreateMonteCarloWorkspace alg;
+    Mantid::HistogramData::HistogramY yData = {1.0, 2.0, 3.0, 4.0};
+    int iterations = alg.computeNumberOfIterations(yData);
+    TS_ASSERT_EQUALS(iterations, 10); // Verify if the sum of yData is rounded correctly
+  }
+
+  void test_computeNormalizedCDF() {
+    CreateMonteCarloWorkspace alg;
+    Mantid::HistogramData::HistogramY yData = {1.0, 2.0, 3.0, 4.0};
+    std::vector<double> cdf = alg.computeNormalizedCDF(yData);
+    TS_ASSERT_EQUALS(cdf.size(), yData.size());
+    TS_ASSERT_DELTA(cdf.back(), 1.0, 1e-6); // Check if the last element is normalized to 1.0
+  }
+
+  void test_fillHistogramWithRandomData() {
+    CreateMonteCarloWorkspace alg;
+    std::vector<double> cdf = {0.1, 0.3, 0.6, 1.0};
+    std::mt19937 gen(32);
+
+    Mantid::HistogramData::HistogramY outputY = alg.fillHistogramWithRandomData(cdf, 100, gen);
+
+    auto sumCounts = std::accumulate(outputY.begin(), outputY.end(), 0.0);
+    TS_ASSERT_EQUALS(sumCounts, 100); // Ensure that the total number of counts is correct
+  }
+
+  void test_fillHistogramWithRandomData_indexOutOfBounds() {
+    CreateMonteCarloWorkspace alg;
+    // Create a CDF where the last value is less than 1.0
+    std::vector<double> cdf = {0.2, 0.5, 0.7, 0.9};
+    std::mt19937 gen(42);
+
+    int numIterations = 1000;
+
+    Mantid::HistogramData::HistogramY outputY = alg.fillHistogramWithRandomData(cdf, numIterations, gen);
+
+    // The total counts should be less than numIterations because some random numbers are not counted
+    auto sumCounts = std::accumulate(outputY.begin(), outputY.end(), 0.0);
+    TS_ASSERT_LESS_THAN(sumCounts, numIterations);
```

Comment:
do you know that this will always happen, is there a chance these two value could be equal?

Reply:
I think these are equal by design in this algorithm, this is in contrast to the method of randomly generating data I posted, where the total counts itself would be Poisson distributed (I believe, I should check...) - I've asked @Despiix to check the distribution of counts in each bin from multiple simulations are Poisson distributed (or close enough).

Reply:
> I think having the parameter to specify number of MC events is good, and the default could be the integral of the input.

@thomashampson - what about making the number of MC events equal to a random number drawn from Poisson distribution with mean of integral of the input? This might help the Poisson-ness of the stats in each bin (if that makes sense...)